src/IB/IBFESurfaceMethod.cpp

```diff
@@ -640,48 +779,389 @@ IBFESurfaceMethod::interpolateVelocity(const int u_data_idx,
+                // via accumulation, i.e., X(qp) = sum_k X_k * phi_k(qp) for
+                // each qp.
+
+                //~
+                for (unsigned int qp = 0; qp < n_qp; ++qp)
+                {
+                    for (unsigned int k = 0; k < NDIM - 1; ++k)
                     {
-                        const double& p = phi[k][qp];
-                        for (unsigned int d = 0; d < NDIM; ++d)
+                        interpolate(dX_dxi[k], qp, X_node, *dphi_dxi[k]);
+                        interpolate(dx_dxi[k], qp, x_node, *dphi_dxi[k]);
+                    }
+                    if (NDIM == 2)
+                    {
+                        dX_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                        dx_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                    }
+                    N = (dX_dxi[0].cross(dX_dxi[1])).unit();
+                    n = (dx_dxi[0].cross(dx_dxi[1])).unit();
+
+                    for (unsigned int i = 0; i < NDIM; ++i)
+                    {
+                        for (unsigned int k = 0; k < n_node; ++k)
                         {
-                            x_qp[NDIM * (qp_offset + qp) + d] += x_node[k][d] * p;
+                            x_qp[NDIM * (qp_offset + qp) + i] += x_node[k][i] * phi[k][qp];
                         }
+                        N_qp[NDIM * (qp_offset + qp) + i] = n(i);
                     }
+
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                        for (unsigned int i = 0; i < NDIM; ++i)
+                            for (unsigned int k = 0; k < n_node; ++k)
+                                DU_j_qp[d][NDIM * (qp_offset + qp) + i] += DU_j_node[d][k][i] * phi[k][qp];
                 }
                 qp_offset += n_qp;
             }
-
-            // Interpolate values from the Cartesian grid patch to the
-            // quadrature points.
-            //
-            // NOTE: Values are interpolated only to those quadrature points
-            // that are within the patch interior.
-            const Box<NDIM>& interp_box = patch->getBox();
-            Pointer<PatchData<NDIM> > u_data = patch->getPatchData(u_data_idx);
+            // Interpolate values from the Cartesian grid patch to the quadrature
+            // points.
+            // Note: Values are interpolated only to those quadrature points that
+            // are within the patch interior
+            const Box<NDIM>& patch_box = patch->getBox();
+            Pointer<PatchData<NDIM> > u_data = patch->getPatchData(side_mask_scratch_idx);
             Pointer<CellData<NDIM, double> > u_cc_data = u_data;
             if (u_cc_data)
             {
                 LEInteractor::interpolate(
-                    U_qp, NDIM, x_qp, NDIM, u_cc_data, patch, interp_box, d_default_interp_spec.kernel_fcn);
+                    U_qp, NDIM, x_qp, NDIM, u_cc_data, patch, patch_box, d_default_interp_spec.kernel_fcn);
```

Comment:
Shouldn't this case result in an error message if velocity jumps are imposed in interpolation?

Reply:
Currently the velocity jump conditions are only setup for staggered grid. The jumps for cell-centered (cc) velocity need to be modified a little bit and can be added later. The current LEInteractor for velocity interpolation ignore the kink but shouldn't result in an error message, or should it?